src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py

```diff
@@ -7888,16 +7892,23 @@ def astype(
             ErrorMessage.not_implemented(
                 f"Snowpark pandas astype API doesn't yet support errors == '{errors}'"
             )
-        col_dtypes_curr = {
-            k: v for k, v in self.dtypes.to_dict().items() if k in col_dtypes_map
-        }
+
+        if is_index:
+            col_dtypes_curr = {
+                column: self.index.dtype for column in self.get_index_names()
+            }
+        else:
+            col_dtypes_curr = {
+                k: v for k, v in self.dtypes.to_dict().items() if k in col_dtypes_map
+            }
 
         astype_mapping = {}
         id_to_sf_type_map = self._modin_frame.quoted_identifier_to_snowflake_type()
         labels = list(col_dtypes_map.keys())
         col_ids = (
             self._modin_frame.get_snowflake_quoted_identifiers_group_by_pandas_labels(
-                labels, include_index=False
+                labels,
```

Comment:
so this relies on the fact that the query compiler for index object only contains index columns, otherwise, it will take both index and data column into account.
That was also why I was suggesting to change name from is_index to include_index. The two set of code added conflicts little bit on the purpose the code above only takes care of index, the implementation here takes both index and data column.

To make the implementation in this function more consistent, you can make the purpose clear to do either one of the following:
1) either apply astype on data column or index column only. If you do this, update get_snowflake_quoted_identifiers_group_by_pandas_labels to also take an parameter include_data_column, and make the function only look into index columns by setting include_data_column = False
2) apply astype on the internalframe with include_index as an option. if you decide to go with this option, then udpate the code for col_dtypes_curr to include both data column and  index column in the dict.

Reply:
I chose to go with option #2 here, what you suggested makes sense. Thanks Yun!!